File: Difficulty/Medium/[143] Reorder List/solution.cpp
/*
 * @lc app=leetcode id=143 lang=cpp
 *
 * [143] Reorder List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

#include <vector>

using namespace std;

class Solution {
public:
    void reorderList(ListNode* head) {
        if (head == nullptr) return;
        vector<ListNode*> vec;
        ListNode *slow = head, *fast = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode *move = slow->next;
        while (move){
            vec.push_back(move);
            move = move->next;
        }
        slow->next = nullptr;
        cout << vec.size() << endl;
        ListNode* prev = head;
        while (!vec.empty()) {
            ListNode* tmp = prev->next;
            ListNode* node_move = vec.back();
            vec.pop_back();
            prev->next = node_move;
            node_move->next = tmp;
            prev = tmp;
        }
    }
};
// @lc code=end

File: Difficulty/Medium/[143] Reorder List/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head: return
        slow, fast, vec = head, head, []
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        move_to_vec = slow.next
        while move_to_vec:
            vec.append(move_to_vec)
            move_to_vec = move_to_vec.next
        slow.next, prev = None, head
        while vec:
            tmp = prev.next
            node_move = vec.pop()
            prev.next = node_move
            node_move.next = tmp
            prev = tmp